### PR TITLE
Dataset unit tests

### PR DESCRIPTION
**Pull request by VasilyBoychuk**
Added docstring comments to Dataset class:
Created Dataset class unit tests:

### CODE REVIEW

File: batchflow/dataset.py [hunk 1]
```diff
@@ -60,15 +142,38 @@ def create_batch(self, batch_indices, pos=False, *args, **kwargs):
...
 
     def __rshift__(self, other):
+        """
+            Parameters
+            ----------
+            other : Pipeline
+
+            Returns
+            -------
+
```
Comment (roman-kh) [inline batchflow/dataset.py]: Missing description

File: batchflow/dataset.py [hunk 2]
```diff
@@ -8,23 +8,65 @@
...
+            index : DatasetIndex or FilesIndex
+                Stores an index for a dataset
+
+            batch_class : Batch or inherited-from-Batch
+                Batch class holds the data and contains processing functions
+
+            preloaded : data-type
+                For smaller dataset it might be convenient to preload all data at once
+                As a result, all created batches will contain a portion of some_data.
+
```
Comment (roman-kh) [inline batchflow/dataset.py]: Empty line is not required (remove it everywhere below)

File: batchflow/dataset.py [hunk 3]
```diff
@@ -8,23 +8,65 @@
...
+            dataset : Dataset
+                Source dataset
+
+            index : DatasetIndex or FilesIndex
+
+            batch_class : Batch or inherited-from-Batch
+
+            Returns
+            -------
+            cls : Dataset
```
Comment (roman-kh) [inline batchflow/dataset.py]: Declare only a data type without a variable name (fix all below as well)

File: batchflow/dataset.py [hunk 4]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
             return index
         return DatasetIndex(index)
 
     @staticmethod
     def _is_same_index(index1, index2):
+        """ Check if index1 and index2 are equals
+
+            Parameters
+            ----------
+            index1 : array-like
```
Comment (roman-kh) [inline batchflow/dataset.py]: Not array-like, it is expected ot be an instance of DatasetIndex or its descendant

File: batchflow/dataset.py [hunk 5]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
+
+            index2 : array-like
+
+            Returns
+            -------
+            The result of two indices comparison
```
Comment (roman-kh) [inline batchflow/dataset.py]: bool

File: batchflow/dataset.py [hunk 6]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
+
+        """
         return (isinstance(index1, type(index2)) or isinstance(index2, type(index1))) and \
                index1.indices.shape == index2.indices.shape and \
                np.all(index1.indices == index2.indices)
 
     def create_subset(self, index):
-        """ Create a dataset based on the given subset of indices """
+        """ Create a dataset based on the given subset of indices
+
```
Comment (roman-kh) [inline batchflow/dataset.py]: Parameters missing

File: batchflow/dataset.py [hunk 7]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
         return type(self).from_dataset(self, index)
 
     def create_batch(self, batch_indices, pos=False, *args, **kwargs):
         """ Create a batch from given indices.
 
+            Parameters
+            ----------
+            batch_indices : DatasetIndex or FilesIndex
+
+            pos : bool
```
Comment (roman-kh) [inline batchflow/dataset.py]: Descripe what ``pos`` is used for

File: batchflow/dataset.py [hunk 8]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
+        """
         return type(self).from_dataset(self, index)
 
     def create_batch(self, batch_indices, pos=False, *args, **kwargs):
         """ Create a batch from given indices.
 
+            Parameters
+            ----------
+            batch_indices : DatasetIndex or FilesIndex
+
```
Comment (roman-kh) [inline batchflow/dataset.py]: A description is missing

File: batchflow/dataset.py [hunk 9]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
+            index : DatasetIndex or Files
+        """
         return type(self).from_dataset(self, index)
 
     def create_batch(self, batch_indices, pos=False, *args, **kwargs):
         """ Create a batch from given indices.
 
+            Parameters
+            ----------
+            batch_indices : DatasetIndex or FilesIndex
```
Comment (roman-kh) [inline batchflow/dataset.py]: Rename ``batch_indices`` to ``index``

File: batchflow/dataset.py [hunk 10]
```diff
@@ -33,24 +75,64 @@ def from_dataset(cls, dataset, index, batch_class=None):
...
+
+            pos : bool
+
+            Returns
+            -------
+            batch : Batch or inherited-from-Batch
```
Comment (roman-kh) [inline batchflow/dataset.py]: No need to mention inheritance as it is implied by default

File: batchflow/dataset.py [hunk 11]
```diff
@@ -60,15 +142,38 @@ def create_batch(self, batch_indices, pos=False, *args, **kwargs):
         return self.batch_class(batch_indices, preloaded=self.preloaded, **kwargs)
 
     def pipeline(self, config=None):
-        """ Start a new data processing workflow """
+        """ Start a new data processing workflow
+
+            Parameters
+            ----------
+            config : Config or dict
+
```
Comment (roman-kh) [inline batchflow/dataset.py]: Describe what config means

File: batchflow/tests/dataset_test.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+import pytest
+import numpy as np
+
+
+from batchflow import Dataset
+
+from batchflow import Baseset
+from batchflow import Batch
+from batchflow import DatasetIndex
+from batchflow import Pipeline
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: The imports section does not comply with PEP8

**Comment by roman-kh** [comment]
Make sure that test code is checked by pylint

File: batchflow/tests/dataset_test.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        pipeline_config = {}
+        new_pipeline = dataset.pipeline(pipeline_config)
+        train_pipeline = (new_pipeline << dataset)
+        assert isinstance(train_pipeline, Pipeline)
+
+    def test_split(self, dataset):
+        assert dataset.train is None
+        dataset.split()
+        assert dataset.train is not None
+        assert dataset.test is not None
```
Comment (dpodvyaznikov) [inline batchflow/tests/dataset_test.py]: Add newline at the end of the file

File: batchflow/tests/dataset_test.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+import pytest
+import numpy as np
+
+from batchflow import Dataset, Batch, ImagesBatch, DatasetIndex, Pipeline
+
+
+@pytest.fixture
+def dataset():
+    index = DatasetIndex(np.arange(100))
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: You don't need `np.arange` here, just 100

File: batchflow/tests/dataset_test.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    return Dataset(index, Batch)
+
+
+class TestDataset:
+    def test_from_dataset(self, dataset):
+        new_index = DatasetIndex(np.arange(25))
+        new_ds = Dataset.from_dataset(dataset, new_index)
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]:
```suggestion
        new_ds = dataset.from_dataset(dataset, new_index)
```

File: batchflow/tests/dataset_test.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        assert isinstance(new_ds, Dataset)
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]:
```suggestion
        assert isinstance(new_ds, dataset.__class__)
```

File: batchflow/tests/dataset_test.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+    def test_rshift(self, dataset):
+        pipeline_config = {}
+        new_pipeline = dataset.pipeline(pipeline_config)
+        train_pipeline = (new_pipeline << dataset)
+        assert isinstance(train_pipeline, Pipeline)
+
+    def test_split(self, dataset):
+        assert dataset.train is None
+        dataset.split()
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: Pass parameters to split to ensure that there will be train and test parts

File: batchflow/tests/dataset_test.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    def test_create_subset_wrong_index(self, dataset):
+        """
+        This method checks for correct execution create_subset method in case of
+        there should be exception during calling this method with index which is not in
+        source dataset index range.
+        Now we raises and handle AssertError
+        """
+        wrong_index = DatasetIndex(np.arange(200, 225))
+        with pytest.raises(AssertionError):
+            new_ds = dataset.create_subset(wrong_index)
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: Perhaps `create_subset` should raise an exception when index is wrong

File: batchflow/dataset.py [hunk 19]
```diff
@@ -130,6 +130,8 @@ def create_subset(self, index):
             -------
             Dataset
         """
+        if not np.isin(index.indices, self.indices).all():
+            raise IndexError
```
Comment (roman-kh) [inline batchflow/dataset.py]: Add some description for a user to understand why the exception was raised

File: batchflow/tests/dataset_test.py [hunk 20]
```diff
@@ -67,6 +61,7 @@ def test_rshift(self, dataset):
 
     def test_split(self, dataset):
         assert dataset.train is None
-        dataset.split()
+        dataset.split((0.8, 0.2))
         assert dataset.train is not None
         assert dataset.test is not None
+        assert dataset.validation is None
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: CHeck that train and test parts have the right length

File: batchflow/tests/dataset_test.py [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+        new_pipeline = dataset.pipeline(pipeline_config)
+        assert isinstance(new_pipeline, Pipeline)
+
+    def test_rshift(self, dataset):
+        pipeline_config = {}
+        new_pipeline = dataset.pipeline(pipeline_config)
+        train_pipeline = (new_pipeline << dataset)
+        assert isinstance(train_pipeline, Pipeline)
+
+    def test_split(self, dataset):
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: This test definitely consists of two cases - split them

File: batchflow/tests/dataset_test.py [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+        assert dataset.train is None
+        train_part, test_part = 0.8, 0.2
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: Avoid universalising tests. Make them more specific - use constants and check the immediate results

File: batchflow/tests/dataset_test.py [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+        dataset.split([train_part, test_part])
+        assert dataset.train is not None
+        assert dataset.test is not None
+        assert dataset.validation is None
+        assert dataset.train.indices.shape[0]/dataset.indices.shape[0] == train_part
```
Comment (roman-kh) [inline batchflow/tests/dataset_test.py]: Keep everything simple if not the simplest

```suggestion
        assert len(dataset.train) == 80
```

**Pull request merged**